README.md
# perceval-finos [![Build Status](https://travis-ci.org/Bitergia/grimoirelab-perceval-finos.svg?branch=master)](https://travis-ci.org/Bitergia/grimoirelab-perceval-finos) [![Coverage Status](https://img.shields.io/coveralls/Bitergia/grimoirelab-perceval-finos.svg)](https://coveralls.io/r/Bitergia/grimoirelab-perceval-finos?branch=master)

Bundle of Perceval backends for FINOS ecosystem.

## Backends

The backends currently managed by this package support the next repositories:

* FINOS meetings attendance

## Requirements

* Python >= 3.4
* python3-requests >= 2.7
* grimoirelab-toolkit >= 0.1.9
* perceval >= 0.12

## Installation

To install this package you will need to clone the repository first:

```
$ git clone https://github.com/Bitergia/grimoirelab-perceval-finos.git
```

Then you can execute the following commands:
```
$ pip3 install -r requirements.txt
$ pip3 install -e .
```

In case you are a developer, you should execute the following commands to install Perceval in your working directory (option `-e`) and the packages of requirements_tests.txt.
```
$ pip3 install -r requirements.txt
$ pip3 install -r requirements_tests.txt
$ pip3 install -e .
```

## Examples

### FINOS meetings

#### Locating via URL

```
$ perceval finosmeetings https://gist.githubusercontent.com/maoo/27db80e0dba349caf95cd3f2af909fe8/raw/d8bdc91ed1ac2c0e5b8e0dc95d5248527c88260d/finos-meetings.csv
```

#### Locating via file path

```
$ perceval finosmeetings file:///Users/m/w/projects/metadata-tool/roster-data.csv
```

## License

Licensed under GNU General Public License (GPL), version 3 or later.


NEWS
# Releases

## perceval-finos 0.1 - (2019-03-15)

** New features and improvements: **

* New set of backends added:
   - FINOS Meetings


tests/base.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2019 Bitergia
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
#
# Authors:
#     Valerio Cosentino <valcos@bitergia.com>
#     Santiago Dueñas <sduenas@bitergia.com>
#

import os
import shutil
import tempfile
import unittest

from perceval.archive import Archive


class TestCaseBackendArchive(unittest.TestCase):
    """Unit tests for Backend using the archive"""

    def setUp(self):
        self.test_path = tempfile.mkdtemp(prefix='perceval-finoss_')
        archive_path = os.path.join(self.test_path, 'myarchive')
        self.archive = Archive.create(archive_path)

    def tearDown(self):
        shutil.rmtree(self.test_path)

    def _test_fetch_from_archive(self, **kwargs):
        """Test whether the method fetch_from_archive works properly"""

        items = [items for items in self.backend_write_archive.fetch(**kwargs)]
        items_archived = [item for item in self.backend_write_archive.fetch_from_archive()]

        self.assertEqual(len(items), len(items_archived))

        for i in range(len(items)):
            item = items[i]
            archived_item = items_archived[i]

            del item['timestamp']
            del archived_item['timestamp']

            self.assertEqual(item, archived_item)


tests/run_tests.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2019 Bitergia
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
#
# Authors:
#     Santiago Dueñas <sduenas@bitergia.com>
#

import sys
import unittest


if __name__ == '__main__':
    test_suite = unittest.TestLoader().discover('.', pattern='test_*.py')
    result = unittest.TextTestRunner(buffer=True).run(test_suite)
    sys.exit(not result.wasSuccessful())


tests/test_finosmeetings.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (C) 2018-2019 Fintech Open Source Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
#
# Authors:
#     Maurizio Pillitu <maoo@finos.org>
#     Santiago Dueñas <sduenas@bitergia.com>
#

import os
import unittest

import httpretty
import pkg_resources

pkg_resources.declare_namespace('perceval.backends')

from perceval.backend import BackendCommandArgumentParser, uuid
from perceval.backends.finos.finosmeetings import (FinosMeetings,
                                                   FinosMeetingsCommand,
                                                   FinosMeetingsClient)


MEETINGS_URL = 'http://example.com/finosmeetings_entries'

requests_http = []


def file_abs_path(filename):
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), filename)


def read_file(filename, mode='r'):
    with open(file_abs_path(filename), mode) as f:
        content = f.read()
    return content


def configure_http_server():
    bodies_entries_job = read_file('data/finosmeetings/finosmeetings_entries.csv')

    http_requests = []

    def request_callback(method, uri, headers):
        last_request = httpretty.last_request()

        if uri.startswith(MEETINGS_URL):
            body = bodies_entries_job
        else:
            body = ''

        requests_http.append(httpretty.last_request())

        http_requests.append(last_request)

        return (200, headers, body)

    httpretty.register_uri(httpretty.GET,
                           MEETINGS_URL,
                           responses=[
                               httpretty.Response(body=request_callback)
                               for _ in range(2)
                           ])

    return http_requests


class TestFinosMeetingsBackend(unittest.TestCase):
    """FinosMeetings backend tests"""

    def test_initialization(self):
        """Test whether attributes are initializated"""

        finosmeetings = FinosMeetings(MEETINGS_URL)

        self.assertTrue(finosmeetings.origin, MEETINGS_URL)
        self.assertIsNone(finosmeetings.client)

    def test_has_archiving(self):
        """Test if it returns True when has_archiving is called"""

        self.assertEqual(FinosMeetings.has_archiving(), False)

    def test_has_resuming(self):
        """Test if it returns False when has_resuming is called"""

        self.assertEqual(FinosMeetings.has_resuming(), False)

    @httpretty.activate
    def test_fetch(self):
        """Test whether a list of entries is returned"""

        http_requests = configure_http_server()

        finosmeetings = FinosMeetings(MEETINGS_URL)

        entries = [entry for entry in finosmeetings.fetch()]
        self.assertEqual(len(entries), 3)
        self.assertEqual(len(http_requests), 1)

        # Test metadata
        expected = [('rob.underwood@finos.org', 'Rob Underwood', 'brooklynrob',
                     'Data Tech', 'Data Tech PMC', '2018-09-28', '2018-09-28T00:00:00+00:00'),
                    ('tosha.ellison@finos.org', 'Tosha Ellison', '', 'Data Tech',
                     'Security Reference Data', '2018-12-11', '2018-12-11T00:00:00+00:00'),
                    ('maoo@finos.org', 'Maurizio Pillitu', 'maoo',
                     'FDC3', 'FDC3 PMC', '2018-10-19', '2018-10-19T00:00:00+00:00')]

        for x in range(len(expected)):
            entry = entries[x]['data']
            self.assertEqual(entries[x]['uuid'], uuid(finosmeetings.origin, FinosMeetings.metadata_id(entry)))
            self.assertEqual(entries[x]['updated_on'], FinosMeetings.metadata_updated_on(entry))
            self.assertEqual(entry['email'], expected[x][0])
            self.assertEqual(entry['name'], expected[x][1])
            self.assertEqual(entry['org'], 'FINOS')
            self.assertEqual(entry['githubid'], expected[x][2])
            self.assertEqual(entry['cm_program'], expected[x][3])
            self.assertEqual(entry['cm_title'], expected[x][4])
            self.assertEqual(entry['date'], expected[x][5])
            self.assertEqual(entry['date_iso_format'], expected[x][6])

    @httpretty.activate
    def test_fetch_from_file(self):
        """Test whether a list of entries is returned from a file definition"""

        finosmeetings = FinosMeetings(
            "file://" + file_abs_path('data/finosmeetings/finosmeetings_entries.csv'))

        entries = [entry for entry in finosmeetings.fetch()]
        self.assertEqual(len(entries), 3)

        # Test metadata
        expected = [('rob.underwood@finos.org', 'Rob Underwood', 'brooklynrob',
                     'Data Tech', 'Data Tech PMC', '2018-09-28', '2018-09-28T00:00:00+00:00'),
                    ('tosha.ellison@finos.org', 'Tosha Ellison', '', 'Data Tech',
                     'Security Reference Data', '2018-12-11', '2018-12-11T00:00:00+00:00'),
                    ('maoo@finos.org', 'Maurizio Pillitu', 'maoo',
                     'FDC3', 'FDC3 PMC', '2018-10-19', '2018-10-19T00:00:00+00:00')]

        for x in range(len(expected)):
            entry = entries[x]['data']
            self.assertEqual(entries[x]['uuid'], uuid(finosmeetings.origin, FinosMeetings.metadata_id(entry)))
            self.assertEqual(entries[x]['updated_on'], FinosMeetings.metadata_updated_on(entry))
            self.assertEqual(entry['email'], expected[x][0])
            self.assertEqual(entry['name'], expected[x][1])
            self.assertEqual(entry['org'], 'FINOS')
            self.assertEqual(entry['githubid'], expected[x][2])
            self.assertEqual(entry['cm_program'], expected[x][3])
            self.assertEqual(entry['cm_title'], expected[x][4])
            self.assertEqual(entry['date'], expected[x][5])
            self.assertEqual(entry['date_iso_format'], expected[x][6])

    @httpretty.activate
    def test_fetch_empty(self):
        """Test whether it works when no entries are fetched"""

        body = """"""
        httpretty.register_uri(httpretty.GET,
                               MEETINGS_URL,
                               body=body, status=200)

        finosmeetings = FinosMeetings(
            MEETINGS_URL)

        entries = [entry for entry in finosmeetings.fetch()]

        self.assertEqual(len(entries), 0)


class TestFinosMeetingsCommand(unittest.TestCase):
    """CSVCommand unit tests"""

    def test_backend_class(self):
        """Test if the backend class is FinosMeetings"""

        self.assertIs(FinosMeetingsCommand.BACKEND, FinosMeetings)

    def test_setup_cmd_parser(self):
        """Test if it parser object is correctly initialized"""

        parser = FinosMeetingsCommand.setup_cmd_parser()
        self.assertIsInstance(parser, BackendCommandArgumentParser)

        args = [MEETINGS_URL]

        parsed_args = parser.parse(*args)
        self.assertEqual(parsed_args.uri, MEETINGS_URL)


class TestFinosMeetingsClient(unittest.TestCase):
    """FinosMeetings API client tests

    These tests not check the body of the response, only if the call
    was well formed and if a response was obtained. Due to this, take
    into account that the body returned on each request might not
    match with the parameters from the request.
    """
    @httpretty.activate
    def test_get_entries(self):
        """Test get_entries API call"""

        # Set up a mock HTTP server
        body = read_file('data/finosmeetings/finosmeetings_entries.csv')
        httpretty.register_uri(httpretty.GET,
                               MEETINGS_URL,
                               body=body, status=200)

        client = FinosMeetingsClient(MEETINGS_URL, ',', True)
        response = client.get_entries()

        self.assertEqual(len(response), 4)


if __name__ == "__main__":
    unittest.main(warnings='ignore')


perceval/backends/finos/finosmeetings.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2018-2019 Fintech Open Source Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
#
# Authors:
#     Maurizio Pillitu <maoo@finos.org>
#     Santiago Dueñas <sduenas@bitergia.com>
#

import csv
import logging
import tempfile

from grimoirelab_toolkit.datetime import (InvalidDateError,
                                          datetime_utcnow,
                                          datetime_to_utc,
                                          str_to_datetime)

from ...backend import (Backend,
                        BackendCommand,
                        BackendCommandArgumentParser)
from ...client import HttpClient

CATEGORY_ENTRY = "finos-meeting-entry"
SEPARATOR = ','
CSV_HEADER = 'email,name,org,githubid,cm_program,cm_title,date'
SKIP_HEADER = True
ID_COLUMNS = 'email,name,date,cm_program,cm_title'
DATE_COLUMN = 'date'
TIMESTAMP = 'timestamp'
DATE_ISO = 'date_iso_format'

logger = logging.getLogger(__name__)


class FinosMeetings(Backend):
    """FinosMeetings backend for Perceval.

    This class retrieves the entries from a FINOS meetings
    attendance table. This table can be stored remotely or
    in a local file.

    To initialize this class the URI to that table must be
    provided. The `uri` will be set as the origin of the data.

    :param uri: URI pointer to FINOS meeting data
    :param tag: label used to mark the data
    :param archive: archive to store/retrieve items
    """
    version = '0.2.0'

    CATEGORIES = [CATEGORY_ENTRY]

    def __init__(self, uri, tag=None, archive=None):
        super().__init__(uri, tag=tag, archive=archive)
        self.client = None

    def fetch(self, category=CATEGORY_ENTRY):
        """Fetch the rows from the meetings table.

        :param category: the category of items to fetch

        :returns: a generator of entries
        """
        kwargs = {}
        items = super().fetch(category, **kwargs)

        return items

    def fetch_items(self, category, **kwargs):
        """Fetch the entries.

        :param category: the category of items to fetch
        :param kwargs: backend arguments

        :returns: a generator of items
        """
        logger.info("Looking for a meeting table at feed '%s'", self.origin)

        nentries = 0
        nskipped = 0

        entries = self.client.get_entries()

        for item in _parse_entries(entries):
            # Need to pass which columns are IDs to metadata_id static function
            ret = {
                '_id_columns': ID_COLUMNS
            }

            for i, column in enumerate(CSV_HEADER.split(',')):
                value = item[i]
                if isinstance(item[i], str):
                    value = item[i].strip()

                # If it's the date column, parse value and add it as 'timestamp' in the item
                if column == DATE_COLUMN:
                    try:
                        dt = str_to_datetime(value)
                        ret[DATE_ISO] = datetime_to_utc(dt).isoformat()
                        ret[TIMESTAMP] = datetime_to_utc(dt).timestamp()
                    except InvalidDateError:
                        logger.warning("Skipping entry due to wrong date format: '%s'", value)
                        nskipped += 1
                        break

                ret[column.strip()] = value

            if 'timestamp' in ret:
                yield ret
                nentries += 1

        logger.info("Done. %s/%s meeting entries fetched; %s ignored",
                    nentries, nentries + nskipped, nskipped)

    @classmethod
    def has_archiving(cls):
        """Returns whether it supports archiving entries on the fetch process.

        :returns: this backend does not support entries archive
        """
        return False

    @classmethod
    def has_resuming(cls):
        """Returns whether it supports to resume the fetch process.

        :returns: this backend does not supports entries resuming
        """
        return False

    @staticmethod
    def metadata_id(item):
        """Extracts the identifier from a meeting item.

        For each entry, the value will be the concatenation
        '_id_columns' values
        """
        string = ""
        for column in item['_id_columns'].split(','):
            string = string + item[column] + "-"
        return string

    @staticmethod
    def metadata_category(item):
        """Extracts the category from a meeting item.

        This backend only generates one type of item which is
        `finos-meetings-entry`.
        """
        return CATEGORY_ENTRY

    @staticmethod
    def metadata_updated_on(item):
        """Extracts the update time from a FINOS meeting CSV row.

        The timestamp is calculated before and stored into the 'timestamp'
        field extracted from 'published' field.

        :param item: item generated by the backend

        :returns: the timestamp field, UNIX format
        """
        return item['timestamp']

    def _init_client(self, from_archive=False):
        """Init client"""

        return FinosMeetingsClient(self.origin, SEPARATOR, SKIP_HEADER)


def _parse_entries(rows):
    ret = []
    for i, row in enumerate(rows):
        if SKIP_HEADER and i == 0:
            logger.debug("skipping header")
        else:
            ret.append(row)
    return ret


class FinosMeetingsClient(HttpClient):
    """FinosMeetings API client.

    :param uri: URI Pointer to CSV contents
    :param archive: an archive to store/read fetched data
    :param from_archive: it tells whether to write/read the archive
    """
    def __init__(self, uri, archive=None, from_archive=False):
        if uri.startswith('file://'):
            self.file_path = uri.split('file://', 1)[1]
        else:
            self.file_path = tempfile.mkdtemp() + "/perceval-finos-meetings-backend-" + str(datetime_utcnow()) + ".csv"
            super().__init__(uri, archive=archive, from_archive=from_archive)
            response = self.session.get(uri)
            open(self.file_path, 'wb').write(response.content)

    def get_entries(self):
        """Retrieve all entries from a CVS file"""

        self.session = None
        with open(self.file_path, newline='') as csv_content:
            reader = csv.reader(csv_content, delimiter=SEPARATOR)
            rows = []
            for row in reader:
                rows.append(row)
            return rows


class FinosMeetingsCommand(BackendCommand):
    """Class to run FinosMeetings backend from the command line."""

    BACKEND = FinosMeetings

    @staticmethod
    def setup_cmd_parser():
        """Returns the FinosMeetings argument parser."""

        parser = BackendCommandArgumentParser()

        # Required arguments
        parser.parser.add_argument('uri',
                                   help="URI pointer to FINOS meetings data")

        return parser
